File: py/consumer.py
import logging
import signal
import sys
from typing import Any

import pika
from pika.adapters.blocking_connection import BlockingConnection

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

QUEUE_NAME = "Server1Queue"
RABBITMQ_URL = "amqp://guest:guest@localhost:5672"


class RabbitMQConsumer:
    def __init__(self):
        self._connection: BlockingConnection = None
        self._channel: pika.adapters.blocking_connection.BlockingChannel = None
        self._should_connect = False
        self._closing = False

    def connect(self) -> None:
        """Extablish RabbitMQ Connection"""
        try:
            self._connection = pika.BlockingConnection(pika.URLParameters(RABBITMQ_URL))
            self._channel = self._connection.channel()
            self._channel.queue_declare(queue=QUEUE_NAME, durable=True)
            logger.info("RabbitMQ connection established")
        except pika.exceptions.AMQPError as e:
            logger.error(f"RabbitMQ connection error: {e}")
            raise

    def start_consuming(self):
        """Start consuming messages from RabbitMQ"""
        try:
            self._channel.basic_consume(
                queue=QUEUE_NAME,
                on_message_callback=self._process_message,
                auto_ack=True,
            )
            logger.info("Waiting for messages...")
            self._channel.start_consuming()
        except pika.exceptions.AMQPError as e:
            logger.error(f"Error while consuming messages: {e}")
            self.reconnect()

    def _process_message(
        self, channel: BlockingConnection, method: Any, properties: Any, body: bytes
    ) -> None:
        """Process incoming messages"""
        try:
            message = body.decode()
            logger.info(f"Received message: {message}")
        except Exception as e:
            logger.error(f"Error processing message: {e}")

    def reconnect(self) -> None:
        """Reconnect to RabbitMQ in not closing"""
        if self._closing:
            return
        self._should_connect = True
        self.stop()

    def stop(self) -> None:
        """Cleanly stop the consumer"""
        if self._channel and self._channel.is_open:
            logger.info("Closing RabbitMQ channel")
            self._channel.close()
        if self._connection and self._connection.is_open:
            logger.info("Closing RabbitMQ connection")
            self._connection.close()
        if self._should_connect:
            logger.info("Reconnecting to RabbitMQ")
            self.connect()
            self.start_consuming()


def main():
    """Main function to run RabbitMQ consumer"""
    consumer = RabbitMQConsumer()
    consumer.connect()

    def shutdown_handler(signum, frame):
        """Handle shutdown signal"""
        logger.info("Received shutdown signal")
        consumer._closing = True
        consumer.stop()
        sys.exit(0)

    signal.signal(signal.SIGINT, shutdown_handler)
    signal.signal(signal.SIGTERM, shutdown_handler)

    try:
        consumer.start_consuming()
    except KeyboardInterrupt:
        shutdown_handler(None, None)


if __name__ == "__main__":
    main()

File: js/consumer.js
import amqp from "amqplib";
import winston from "winston";

// Configure logging
const logger = winston.createLogger({
  level: "info",
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.json(),
  ),
  transports: [new winston.transports.Console()],
});

const QUEUE_NAME = "Server1Queue";
const RABBITMQ_URL = "amqp://guest:guest@localhost:5672";

class RabbitMQConsumer {
  constructor() {
    this.connection = null;
    this.channel = null;
  }

  async connect() {
    try {
      this.connection = await amqp.connect(RABBITMQ_URL);
      this.channel = await this.connection.createChannel();

      await this.channel.assertQueue(QUEUE_NAME, {
        durable: true,
      });

      logger.info("RabbitMQ connection established");
    } catch (err) {
      logger.error(`RabbitMQ connection failed: ${err}`);
      throw err;
    }
  }

  async startConsuming() {
    try {
      await this.channel.consume(QUEUE_NAME, (message) => {
        if (message !== null) {
          try {
            const content = message.content.toString();
            logger.info(`Received message: ${content}`);

            // Process message here
            this.processMessage(content);

            this.channel.ack(message);
          } catch (err) {
            logger.error(`Message processing failed: ${err}`);
            this.channel.nack(message);
          }
        }
      });

      logger.info(`Waiting for messages in queue: ${QUEUE_NAME}`);
    } catch (err) {
      logger.error(`Consuming failed: ${err}`);
      throw err;
    }
  }

  processMessage(content) {
    // Implement your message processing logic
    logger.info(`Processing message: ${content}`);
  }

  async close() {
    try {
      if (this.channel) await this.channel.close();
      if (this.connection) await this.connection.close();
      logger.info("RabbitMQ connection closed");
    } catch (err) {
      logger.error(`Error closing connection: ${err}`);
    }
  }
}

// Main execution
const consumer = new RabbitMQConsumer();

async function main() {
  try {
    await consumer.connect();
    await consumer.startConsuming();

    // Graceful shutdown
    process.on("SIGTERM", async () => {
      logger.info("SIGTERM received - shutting down");
      await consumer.close();
      process.exit(0);
    });

    process.on("SIGINT", async () => {
      logger.info("SIGINT received - shutting down");
      await consumer.close();
      process.exit(0);
    });
  } catch (err) {
    logger.error(`Consumer failed: ${err}`);
    process.exit(1);
  }
}

main();
